### DISCUSSION TITLE
Typing JSON field selects

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [romeerez/orchid-orm#113] [Q&A] [answered]
How to type JSON field sub-select? Consider the table:

```ts
export class Table extends BaseTable {
  readonly table = 'table'
  columns = this.setColumns(t => ({
    data: t.json(t =>
      t.object({
        foo: t.record(t.string(), t.object({ bar: t.string() })),
      }),
    ),
  }))
}

export type TableDataFoo = TableType<Table>['data']['foo']
```

and the query:

```ts
const row = await db.table.take().select({ foo: db.table.sql`data->'foo'` })
row.foo // unknown
```

I know I can use `db.table.sql(t => t.whatever)` but that involves copy/pasting the entire JSON subtree:

```ts
const row = await db.table.take().select({
  foo: db.table.sql(t =>
    // But what if this definition is like 30x times larger?
    t.json(t => t.record(t.string(), t.object({ bar: t.string() }))),
  )`data->'foo'`,
})
row.foo // Record<string, { bar: string }>
```

Is there some recipe for streamlining JSON subfield queries?

**Comment by IlyaSemenov** [comment]
Just an idea, what if this was possible?

```ts
const row = await db.table.take().select({
  foo: db.table.sql`data->'foo'`.castTo<TableDataFoo>(),
})
```

(Naming taken from [Objection.js](https://vincit.github.io/objection.js/api/types/#castto)).

This could be useful for other raw SQL as well.

**Reply by IlyaSemenov** [reply, marked as answer]
This was implemented in a similar fashion: https://github.com/romeerez/orchid-orm/pull/118#issuecomment-1605834724 🎉

**Comment by romeerez** [comment]
The easies, but not type-safe, would be to select with a raw SQL and cast unknown to your type later:

```ts
const row = await db.table.take().select({
  foo: db.table.sql`data->'foo'`,
})
const data = row.foo as YourType
```

Postgres offers a lot of different functionality to work with JSON, and I planned to dive deeper into this sometime later and find ways to work with it with type safety. For now, let's not add functionality for unsafe type casting, it can be done with native TS `as` anyway.
